oem/firstvoices/android/app/src/main/java/com/firstvoices/keyboards/FVShared.java

```diff
@@ -124,53 +129,83 @@ public static FVShared getInstance() {
-
-                    region.keyboards.add(keyboard);
-                }
+      FVRegionList list = new FVRegionList();
+      File resourceRoot = new File(getResourceRoot());
+      PackageProcessor kmpProcessor = new PackageProcessor(resourceRoot);
+      JSONParser parser = new JSONParser();
+      File jsonFile = new File(getPackagesDir() + FVDefault_PackageID + File.separator + FVKeyboards_JSON);
+      if (!jsonFile.exists()) {
+          return list;
+      }
+      try {
+          // At this point in initialization, fv_all.kmp is now extracted, so
+          // populate keyboard info from keyboards.json
+          JSONArray keyboardsArray = parser.getJSONObjectFromFile(jsonFile, JSONArray.class);
+
+          if (keyboardsArray == null) {
+              Log.d("loadRegionList", "unable to load keyboards.json");
+              return list;
+          }
 
-                line = reader.readLine();
-            }
+          for (int i=0; i<keyboardsArray.length(); i++) {
+              JSONObject keyboardObj = keyboardsArray.getJSONObject(i);
+              if (keyboardObj == null) {
+                  Log.d("loadRegionList", "keyboard Obj is null");
+                  continue;
+              }
+              JSONArray languageArray = keyboardObj.getJSONArray("languages");
+              if (languageArray == null) {
+                  Log.d("loadRegionList", "language Array is null");
+                  continue;
+              }
+              String regionName = keyboardObj.getString("region");
+              FVRegion region = list.findRegion(regionName);
+              if(region == null) {
+                  region = new FVRegion(regionName);
+                  list.add(region);
+              }
+
+              String kbID = keyboardObj.getString("id");
+              String kbName = keyboardObj.getString("name");
+              JSONObject languageObj = languageArray.getJSONObject(0);
+              String lgID = languageObj.getString("id").toLowerCase(); // Normalize language Id
+              String lgName = languageObj.getString("name");
+
+              // Override European keyboard info
+              if (kbID.equalsIgnoreCase("sil_euro_latin")) {
+                kbName = "English";
+                lgID = "en";
+                lgName = "English";
+              } else if (kbID.equalsIgnoreCase("basic_kbdcan")) {
+                kbName = "Français";
+                lgID = "fr-ca";
+                lgName = "French";
+              }
+
+              FVKeyboard kbd = new FVKeyboard(
+                  kbID,
+                  kbName,
+                  kbID, // unused legacy ID
+                  keyboardObj.getString("version"),
+                  lgID,
+                  lgName);
+
+              region.keyboards.add(kbd);
+          }
+      } catch (Exception e) {
+          Log.e("createKeyboardList", "Error: " + e);
+          // We'll return a malformed list for now in this situation
+      }
```

Comment:
We should be crashing the app so we can detect this -- a malformed list will make the app unusable anyway, won't it?

Reply:
This block was pre-existing.
I've updated to throw fatal error